[nnc] Fixed checking for loop carried dependence while fusing 2D reduction loops

navahgar opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#60609**

Fixes #60310

Differential Revision: [D29386144](https://our.internmc.facebook.com/intern/diff/D29386144)

navahgar committed:
[nnc] Fixed checking for loop carried dependence while fusing 2D reduction loops

[ghstack-poisoned]

navahgar committed:
Update on "[nnc] Fixed checking for loop carried dependence while fusing 2D reduction loops"


Fixes #60310 

[ghstack-poisoned]

navahgar commented:
@navahgar has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29386144).

huiguoo commented on test/cpp/tensorexpr/test_loopnest.cpp:
```diff
@@ -5868,6 +5868,163 @@ TEST(LoopNest, fuseLoopsWithReductions) {
+  // Input IR:
+  //   for (int i = 0; i < 20; i++) {
+  //     for (int j = 0; j < 20; j++) {
+  //       A[i,j*20] = i + j;
```
A[i, j*20] -> A[i, j*20+i]

navahgar replied:
Fixed

huiguoo commented on test/cpp/tensorexpr/test_loopnest.cpp:
```diff
@@ -5868,6 +5868,163 @@ TEST(LoopNest, fuseLoopsWithReductions) {
+  //   }
+  //   for (int m = 0; m < 20; m++) {
+  //     for (int n = 0; n < 20; n++) {
+  //       B[m,n] = A[m,n*20];  // Both indices of A use m
```
A[m, n * 20] -> A[m, m * 20 + n]

navahgar replied:
Fixed.

huiguoo commented on test/cpp/tensorexpr/test_loopnest.cpp:
```diff
@@ -5868,6 +5868,163 @@ TEST(LoopNest, fuseLoopsWithReductions) {
+
+  // NOLINTNEXTLINE(cppcoreguidelines-init-variables)
+  For* fused_loop;
+  ASSERT_FALSE(LoopNest::fuseLoops({forI, forM}, &fused_loop));
```
This case is not handled by the `equivalence` check but the indices here are equivalent, right? That is, the write and read of `A[m, m*20+n]` happens in the same iteration of the outer loop?

navahgar replied:
Yes, this is one of the case where the fusion will not be allowed because we cannot prove the accesses to `A` are loop-independent after fusion.

huiguoo commented on test/cpp/tensorexpr/test_loopnest.cpp:
```diff
@@ -5868,6 +5868,163 @@ TEST(LoopNest, fuseLoopsWithReductions) {
+}
+
+// NOLINTNEXTLINE(cppcoreguidelines-avoid-non-const-global-variables)
+TEST(LoopNest, fuseLoopsWithMixedLoopVarsAsIndices) {
```
Same test as above?

navahgar replied:
Good catch. I had some local changes which I missed pushing here. These are actually 2 different tests, one where the loop indices are complex but the outer loop var is not used in inner dim and other where the outer loop var is also used in the inner dim.

navahgar committed:
Update on "[nnc] Fixed checking for loop carried dependence while fusing 2D reduction loops"


Fixes #60310

Differential Revision: [D29386144](https://our.internmc.facebook.com/intern/diff/D29386144)

[ghstack-poisoned]

navahgar commented:
@navahgar has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29386144).

bertmaher commented on test/cpp/tensorexpr/test_loopnest.cpp:
```diff
@@ -5868,6 +5868,180 @@ TEST(LoopNest, fuseLoopsWithReductions) {
+  VarHandle k("k", kInt);
+  VarHandle m("m", kInt);
+  VarHandle n("n", kInt);
+  auto initA = Store::make(a_buf, {i, j}, 0);
+  auto sumA = Store::make(
+      a_buf,
+      {i, j},
+      Add::make(Load::make(a_buf, {i, j}), Load::make(b_buf, {i, j, k})));
+  auto forK = For::make(k, 0, 100, sumA);
+  auto forJ = For::make(j, 0, 50, Block::make({initA, forK}));
+  auto forI = For::make(i, 0, 20, forJ);
+  auto storeC = Store::make(c_buf, {m, n}, Load::make(a_buf, {m, n}));
+  auto forM = For::make(m, 0, 20, For::make(n, 0, 40, storeC));
+  auto par = Block::make({forI, forM});
```
Is there a reason to avoid using `Compute` in these tests?

navahgar replied:
Just personal preference I guess. In the past, I have had some cases where `Compute` wasn't enough to write some tests. So, I resorted to generating the IR from scratch. It helps me understand the IR better. Also, in some cases I need to refer to some intermediate loop or statements later in the test (for some checks). I can have a handle to those when I generate the tests this way.

bertmaher reviewed (approved):
I was probably being dense, but it took me a long time to understand why it's acceptable to completely ignore indices that don't involve the outer-loop variables.  If I understand correctly, it's because fusion (which changes the outer loop variables) won't change the ordering of accesses with respect to those dimensions.

I think I might be thrown by the wording of "equivalence".  Like if you're fusing loop `i`, and you're considering indices `j * 2` vs `k * 3`, those expressions don't look "equivalent" to me, but we just don't care about them.

If there's some better naming here that could help clarify it'd be great (but no need to block on this (unless the above indicates that in fact I *don't* understand the issue, in which case more explanation is probably needed :-) )).

navahgar committed:
Update on "[nnc] Fixed checking for loop carried dependence while fusing 2D reduction loops"


Fixes #60310

Differential Revision: [D29386144](https://our.internmc.facebook.com/intern/diff/D29386144)

[ghstack-poisoned]

navahgar commented:
@navahgar has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29386144).

navahgar commented:
>  If I understand correctly, it's because fusion (which changes the outer loop variables) won't change the ordering of accesses with respect to those dimensions.

You are right about this.

> I think I might be thrown by the wording of "equivalence".

Yeah,  I realize that now. I have renamed the function and also updated the comments to explain things better now. Hopefully, this is more clear :). 

Thanks for pointing that out.

navahgar committed:
Update on "[nnc] Fixed checking for loop carried dependence while fusing 2D reduction loops"


Fixes #60310

Differential Revision: [D29386144](https://our.internmc.facebook.com/intern/diff/D29386144)

[ghstack-poisoned]

navahgar commented:
@navahgar has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29386144).
